src/PowerShellEditorServices/Services/DebugAdapter/Handlers/StackTraceHandler.cs

```diff
@@ -1,64 +1,132 @@
-                maxFrameCount = Math.Min(maxFrameCount, startFrameIndex + requestedFrameCount);
-            }
+        StackFrame breakpointLabel = CreateBreakpointLabel(invocationInfo);
 
-            for (long i = startFrameIndex; i < maxFrameCount; i++)
+        if (skip == 0 && take == 1) // This indicates the client is doing an initial fetch, so we want to return quickly to unblock the UI and wait on the remaining stack frames for the subsequent requests.
+        {
+            return new StackTraceResponse()
             {
-                // Create the new StackFrame object with an ID that can
-                // be referenced back to the current list of stack frames
-                newStackFrames.Add(LspDebugUtils.CreateStackFrame(stackFrameDetails[i], id: i));
-            }
+                StackFrames = new StackFrame[] { breakpointLabel },
+                TotalFrames = INITIAL_PAGE_SIZE //Indicate to the client that there are more frames available
+            };
+        }
+
+        // Wait until the stack frames and variables have been fetched.
+        await debugService.StackFramesAndVariablesFetched.ConfigureAwait(false);
+
+        StackFrameDetails[] stackFrameDetails = await debugService.GetStackFramesAsync(cancellationToken)
+                                                                    .ConfigureAwait(false);
 
+        // Handle a rare race condition where the adapter requests stack frames before they've
+        // begun building.
+        if (stackFrameDetails is null)
+        {
             return new StackTraceResponse
             {
-                StackFrames = newStackFrames,
-                TotalFrames = newStackFrames.Count
+                StackFrames = Array.Empty<StackFrame>(),
+                TotalFrames = 0
+            };
+        }
+
+        List<StackFrame> newStackFrames = new();
+        if (skip == 0)
+        {
+            newStackFrames.Add(breakpointLabel);
+        }
+
+        newStackFrames.AddRange(
+            stackFrameDetails
+            .Skip(skip != 0 ? skip - 1 : skip)
+            .Take(take != 0 ? take - 1 : take)
+            .Select((frame, index) => CreateStackFrame(frame, index + 1))
+        );
+
+        return new StackTraceResponse
+        {
+            StackFrames = newStackFrames,
+            TotalFrames = newStackFrames.Count
+        };
+    }
+
+    public static StackFrame CreateStackFrame(StackFrameDetails stackFrame, long id)
+    {
+        SourcePresentationHint sourcePresentationHint =
+            stackFrame.IsExternalCode ? SourcePresentationHint.Deemphasize : SourcePresentationHint.Normal;
+
+        // When debugging an interactive session, the ScriptPath is <No File> which is not a valid source file.
+        // We need to make sure the user can't open the file associated with this stack frame.
+        // It will generate a VSCode error in this case.
+        Source? source = null;
+        if (!stackFrame.ScriptPath.Contains("<"))
```

Comment:
So, I get you're checking for `<No File>` which is fair _but_ `<` is a valid file character on non-Windows (err NTFS) filesystems, so you need to check for the whole thing `<No File>` (and we might also need to check for translations of that? Unsure...and while important, we're already not handling localization super great).

Reply:
Good point, will do.

Reply:
I think I meant for it to be specifically that and maybe I got off on another idea or something. Good Catch.